reps/2024-05-21-kuberay-authentication.md

```diff
@@ -0,0 +1,119 @@
+# Kubernetes Native Ray Authentication
+
+Ray, in its default configuration, lacks robust security and authentication measures.
+Its deployment on Kubernetes provides some level of protection by leveraging RBAC for access control of RayCluster resources.
+However, once provisioned, a RayCluster remains vulnerable to unauthorized access from anyone with network connectivity to the Ray head node.
+
+This proposal introduces a Kubernetes aware sidecar proxy for authenticating external access to the Ray head.
+It will leverage the existing Kubernetes authentication system to grant tokens used to securely access Ray head endpoints.
+This simplifies Ray authentication by centralizing management of tokens and users within Kubernetes.
+
+## Authentication Scope
+
+While there is a large surface area of a Ray cluster that could benefit from authenticated access, this proposal focuses specifically on securing Ray head endpoints that are frequently accessed externally.
+This includes enforcing authentication for both the dashboard server and the interactive client server. Securing access from internal clients like the Raylet to the GCS server is not addressed in this proposal
+as network policies are typically sufficient to protect this communication.
+
+## Sidecar Authentication Proxy
+
+To enforce authenticated access to specific Ray head ports, the KubeRay operator will deploy a sidecar container alongside the Ray head pod.
+This sidecar will function as a reverse proxy, validating authentication tokens for requests to the dashboard port (8265) and the interactive client port (10001).
+Traffic to other ports will pass through the sidecar proxy without requiring authentication.
+
+The KubeRay operator will be changed in the following ways:
+1. The Ray head container will bind to localhost only
+2. A reverse proxy sidecar will run alongside the Ray head container
+3. A ServiceAccount is created per RayCluster resource
+
+The sidecar ensures only requests with the correct authorization tokens can access the Ray head. More details on how tokens are authenticated below.
+For the MVP, the sidecar will be implemented using Envoy configured with an [External Authorizer](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter).
+
+## Access Control with TokenReview API
+
+The sidecar authenticator will use the TokenReview API to authenticate tokens passed into the authorization headers of each request.
+Tokens can be created from Kubernetes ServiceAccounts or an external identity provider as long as the Kubernetes API Server is configured with an external authorizer.
+Each RayCluster will be provisioned a default ServiceAccount that KubeRay Operator will use when authenticating to the Ray head (specifically for RayJob and RayService).
```

Comment:
It may be useful for the implementation phase, to explicitly mention here the case of RayJobs with submission mode set to `K8sJobMode`, for which the Ray CLI command executed in the submission Job will have to be authenticated as well.

Reply:
Good point, will add that!